### PR TITLE
Support S3 URI using aws-sdk-cpp libraries. Also setup minio for testing with github actions.

### PR DESCRIPTION
**Pull request by nalinigans**
This is a first pass at supporting S3 URIs with [aws-sdk-cpp](https://github.com/aws/aws-sdk-cpp.git). 
1. The uri supported should be of the form `s3://bucket_name/folder1/file`.
2. Authentication at this point is either via aws specific environment variables and/or from the config/credential files in $HOME/.aws. Also introduced a new environment variable for endpoint overrides - `AWS_ENDPOINT_OVERRIDE` as support for one does not exist in the sdk. We could extend the uri with query parameters as necessary.
3. The upload/download support is currently synchronous, will start work on async operations soon.
4. Testing was done on aws s3 locally, but is done with Minio/s3 for GitHub workflows. Minio has its own quirks, we will have to start treating s3/azure/gcs storage as object stores rather than virtual filesystems as much as possible. Basically, move away from operations like create_dir and is_dir.

### CODE REVIEW

**Review by mlathara — APPROVED** [review]
looks good. just some minor clarifying questions.

File: core/src/storage_manager/storage_s3.cc [hunk 1]
```diff
@@ -0,0 +1,489 @@
...
+      auto update_info = multipart_upload_info_t(upload_id);
+      write_map_.insert({filepath, std::move(update_info)});
+    }
+    auto found = write_map_.find(filepath);
+    assert(found != write_map_.end());
+    found->second.part_number_++;
+    upload_id = found->second.upload_id_;
+    part_number = found->second.part_number_;
+    completed_parts = found->second.completed_parts_;
+    // Verify that the previous uploaded part was at least 5M - see https://docs.aws.amazon.com/AmazonS3/latest/API/mpUploadUploadPart.html
```
[the remark this answers is not in the document]
Comment (nalinigans) [reply core/src/storage_manager/storage_s3.cc:396]: Yep, good point maybe we should warn and set it to 5M.

File: core/src/storage_manager/storage_s3.cc [hunk 2]
```diff
@@ -0,0 +1,489 @@
...
+  Aws::S3::Model::ListObjectsV2Request request;
+  request.SetBucket(bucket_name_);
+  request.SetPrefix(path);
+  request.SetDelimiter(to_aws_string(DELIMITER));
+  Aws::String continuation_token = "";
+  do {
+    if (continuation_token.length() > 0) {
+      request.SetContinuationToken(continuation_token);
+    }
+    auto outcome = client_->ListObjectsV2(request);
```
[the remark this answers is not in the document]
Comment (nalinigans) [reply core/src/storage_manager/storage_s3.cc:294]: I will add the check, but even as is works as the result will not return anything.

File: core/src/storage_manager/storage_s3.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (nalinigans) [reply core/src/storage_manager/storage_s3.cc:400]: Changed to put out a better error message - https://github.com/OmicsDataAutomation/TileDB/blob/0c3ba3b3e33fa8b517d5d0b8caa7c2e48e0a8ea0/core/src/storage_manager/storage_s3.cc#L403. We can refine this later.

File: core/src/storage_manager/storage_s3.cc [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
```
[the remark this answers is not in the document]
Comment (nalinigans) [reply core/src/storage_manager/storage_s3.cc:273]: Done.

File: core/src/storage_manager/storage_s3.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (nalinigans) [reply core/src/storage_manager/storage_s3.cc:296]: Done.

**Review by mlathara — APPROVED** [review]
looks good

**Pull request merged**